src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,2919 @@
+            return;
+        }
+
+        lanes = op->type.lanes();
+    }
+
+    void visit(const Store *op) override {
+        IRVisitor::visit(op);
+
+        if (op->name != buffer_name) {
+            return;
+        }
+
+        if (op->value.type().is_scalar()) {
+            are_all_dense = false;
+            return;
+        }
+
+        Expr ramp_base = strided_ramp_base(op->index);
+        if (!ramp_base.defined()) {
+            are_all_dense = false;
+            return;
+        }
+
+        if ((op->alignment.modulus % op->value.type().lanes() != 0) ||
+            (op->alignment.remainder % op->value.type().lanes() != 0)) {
+            are_all_dense = false;
+            return;
+        }
+
+        if (lanes != -1 && op->value.type().lanes() != lanes) {
+            are_all_dense = false;
+            return;
+        }
+
+        lanes = op->value.type().lanes();
+    }
+};
+
+struct FindWorkGroupSize : public IRVisitor {
+    using IRVisitor::visit;
+    void visit(const For *loop) override {
+        if (!CodeGen_GPU_Dev::is_gpu_var(loop->name)) {
+            return loop->body.accept(this);
+        }
+
+        if ((loop->for_type == ForType::GPUBlock) ||
+            (loop->for_type == ForType::GPUThread)) {
+
+            // This should always be true at this point in codegen
+            internal_assert(is_const_zero(loop->min));
+
+            // Save & validate the workgroup size
+            int index = thread_loop_workgroup_index(loop->name);
+            if (index >= 0) {
+                const IntImm *literal = loop->extent.as<IntImm>();
+                if (literal != nullptr) {
+                    uint32_t new_wg_size = literal->value;
+                    user_assert(workgroup_size[index] == 0 || workgroup_size[index] == new_wg_size)
+                        << "Vulkan requires all kernels have the same workgroup size, "
+                        << "but two different sizes were encountered: "
+                        << workgroup_size[index] << " and "
+                        << new_wg_size << " in dimension " << index << "\n";
+                    workgroup_size[index] = new_wg_size;
+                }
+            }
+            debug(4) << "Thread group size for index " << index << " is " << workgroup_size[index] << "\n";
+        }
+        loop->body.accept(this);
+    }
+
+    int thread_loop_workgroup_index(const std::string &name) {
+        std::string ids[] = {".__thread_id_x",
+                             ".__thread_id_y",
+                             ".__thread_id_z"};
+        for (size_t i = 0; i < sizeof(ids) / sizeof(std::string); i++) {
+            if (ends_with(name, ids[i])) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    uint32_t workgroup_size[3] = {0, 0, 0};
+};
+
+CodeGen_Vulkan_Dev::SPIRV_Emitter::SPIRV_Emitter(Target t)
+    : IRVisitor(), target(t) {
+    // Empty
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::scalarize(const Expr &e) {
```

Comment:
Really feels like this should be an IR->IR pass somewhere earlier in the pipeline.  Can we create an issue for that?  I imagine this or similar code appears in multiple GPU backends.

Reply:
Completely agree!  I'll add an issue for this!

Reply:
Added issue ... tracking this as:

Refactor and consolidate common GPU backend IR operations #7238